### ISSUE TITLE
Detect fake markdown links

### ISSUE THREAD

**Issue by Martoon-00** [serokell/xrefcheck#79]
# Clarification and motivation

Quite a common typo is to write `[text][link]` instead of `[text](link)`. The former is also a valid markdown syntax (footnotes); but when it is used mistakenly, it is quite probable that there is no respective footnote, so we could detect that.

I'm not sure whether this is easy to do with our markdown parser, AFAIR it tries to substitute footnotes on the fly. We should investigate this.

# Acceptance criteria

* In case someone used `[text][link]` and `link` is not a footnote of this current document, we should report a respective error.

**Comment by dcastro** [comment]
> I'm not sure whether this is easy to do with our markdown parser, AFAIR it tries to substitute footnotes on the fly. We should investigate this.

Just looked into this a little bit.

If we try to parse this markdown file:

```md
Some [text][b]

 [b]: http://www.google.com
```

The `cmark-gfm` package will return this (printed with the `pretty-simple` package):

```hs
Node                           
    ( Just
        ( PosInfo ...
        )
    ) DOCUMENT
    [ Node
        ( Just
            ( PosInfo ...
            )
        ) PARAGRAPH
        [ Node
            ( Just
                ( PosInfo ...
                )
            )
            ( TEXT "Some " ) []
        , Node
            ( Just
                ( PosInfo ...
                )
            )
            ( LINK "http://www.google.com" "" )
            [ Node
                ( Just
                    ( PosInfo ...
                    )
                )
                ( TEXT "text" ) []
            ]
        ]
    ]

```

As you can see, there's only one node with the text "a" and the link to "google.com". So yes, the parser substitutes footnotes on the fly.

On the other hand, if we try parsing this file with a broken footnote:

```md
Some [text][c]

 [b]: http://www.google.com
```

```hs
Node                           
    ( Just
        ( PosInfo ...
        )
    ) DOCUMENT
    [ Node
        ( Just
            ( PosInfo ...
            )
        ) PARAGRAPH
        [ Node
            ( Just
                ( PosInfo ...
                )
            )
            ( TEXT "Some [text][c]" ) []
        ]
    ]

```

It simply treats it as a text node.

**Comment by dcastro** [comment]
> In case someone used [text][link] and link is not a footnote of this current document, we should report a respective error.

I agree. If someone types `[a][b]`, it's **very** likely that they meant to link to a footnote.

It's also possible that they meant that to be treated as text. When that happens, the user can always use a `<!-- xrefcheck: ignore link -->` annotation.

**Comment by dcastro** [comment]
> In case someone used [text][link] and link is not a footnote of this current document, we should report a respective error.

I agree. If someone types `[a][b]`, it's **very** likely that they meant to link to a footnote.

----

But it's also possible that they meant for it to be treated as plain text, we can't rule that out.

In this scenario, I'd say the user would use a `<!--xrefcheck: ignore <mode> -->` annotation.

Let's write bats tests to make sure that this new "footnote check" and the existing annotations work well together.

**Comment by Sorokin-Anton** [comment]
Actually, `[a][b]` sytnax is not for footnotes but for reusing links (so you can define a link in bottom of document and then use in multiple places), but I agree that if after cmark-gfm there is a text block with `[a][b]` we should report this (currently working on it). 
We also can report broken footnotes (you can read about footnote syntax [here](https://github.blog/changelog/2021-09-30-footnotes-now-supported-in-markdown-fields/)), for example `this[^1]` is a broken footnote if there is no `[^1]: lalala` in bottom of document. @dcastro I suggest to report `[a][b]`, `a[^b]`, but not report `a[b]` or `a [b]`

**Comment by Sorokin-Anton** [comment]
Also currently we're not reporting broken links declared atbottom of md file, if they are unused. For example, document containing only `[unused-footnote-aaa]: aaa` is empty for `cmark-gfm`. Such things are invisible in any renderer, so I hink we should not care about it

**Comment by dcastro** [comment]
> Actually, `[a][b]` sytnax is not for footnotes

Ah yes, my bad, I got my terminology mixed up! It seems I should have said ["reference links"](https://github.github.com/gfm/#reference-link). I've updated the issue title.


There are 3 types of reference links:

1. [full reference link](https://github.github.com/gfm/#full-reference-link)
    ```
    [foo][bar]

    [bar]: /url "title"
    ```
2. [collapsed reference link](https://github.github.com/gfm/#collapsed-reference-link)
    ```
    [foo][]

    [foo]: /url "title"
    ```
3. [shortcut reference link](https://github.github.com/gfm/#shortcut-reference-link)
    ```
    [foo]

    [foo]: /url "title"
    ```

Note that the shortcut reference link `[foo]` and the collapsed reference link `[foo][]` should be interpreted the same as the full reference link `[foo][foo]`.

-----

However, we have a bigger problem here, and I think it may be a showstopper.

As I mentioned in [an earlier comment](https://github.com/serokell/xrefcheck/issues/79#issuecomment-1243003572), it is possible the user may `[a]` to be interpreted as a literal string, rather than a reference link.

The correct way to do that is to escape the brackets, i.e.:

```md
<!-- This should be interpreted as a "shortcut reference link", 
     IFF a "link reference definition" matching "a" exists. 
-->
[a]

<!-- Interpret as a literal string and render the characters '[', 'a', ']' -->
\[a\]
```

The problem is that, when a "link reference definition" matching "a" does NOT exist, `cmark-gfm` parses the two nodes exactly the same:

```hs
    [ Node
        ( Just
            ( PosInfo
                { startLine = 1
                , startColumn = 1
                , endLine = 1
                , endColumn = 5
                }
            )
        ) PARAGRAPH
        [ Node
            ( Just
                ( PosInfo
                    { startLine = 1
                    , startColumn = 1
                    , endLine = 1
                    , endColumn = 5
                    }
                )
            )
            ( TEXT "[a]" ) []
        ]
    ]
```

If we install [GitHub's official `cmark-gfm` binary](https://github.com/github/cmark-gfm#installing), we'll observe the same behavior:

```
$ printf '[a]' | cmark-gfm -t xml 

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE document SYSTEM "CommonMark.dtd">
<document xmlns="http://commonmark.org/xml/1.0">
  <paragraph>
    <text xml:space="preserve">[a]</text>
  </paragraph>
</document>

$ printf '\[a\]' | cmark-gfm -t xml         

<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE document SYSTEM "CommonMark.dtd">
<document xmlns="http://commonmark.org/xml/1.0">
  <paragraph>
    <text xml:space="preserve">[a]</text>
  </paragraph>
</document>
```

This means that it's impossible for us to know whether the user wants this `Node` AST to be interpreted as a literal or as a shortcut reference link.

Similarly, `[^1]` and `\[^1\]` are also parsed into the same AST.

----

In conclusion, I don't think this issue can be solved in a spec-compliant way :/
